bitfield/impl/src/bitfield.rs

```diff
@@ -0,0 +1,245 @@
+                    let new = input.pop_bits(8 - lsb_offset as u32);
+                    self.data[ls_byte] = stays_same | (new << lsb_offset as u32);
+
+                    if ms_byte - ls_byte >= 2 {
+                        // Middle bytes
+                        for byte in self.data[(ls_byte + 1)..ms_byte].iter_mut() {
+                            *byte = input.pop_bits(8);
+                        }
+                    }
+
+                    if ls_byte != ms_byte {
+                        // Most-significant byte
+                        self.data[ms_byte] |= input.pop_bits(msb_offset as u32);
+                    }
+                }
+            });
+
+            let mut offset = Punctuated::<syn::Expr, Token![+]>::new();
+            offset.push(syn::parse_quote!{ 0 });
+            for field in fields_named.named.iter() {
+                use crate::ident_ext::IdentExt as _;
+                let field_name = field.ident.clone().expect("named fields is already guaranteed; qed");
+                let getter_name = syn::Ident::from_str(&format!("get_{}", field_name));
+                let setter_name = syn::Ident::from_str(&format!("set_{}", field_name));
+                let field_type = &field.ty;
+
+                let mut bits_check_tokens = quote! {};
+                for attr in field.attrs.iter().filter(|attr| attr.path.is_ident("bits")) {
+                    let bits_arg = syn::parse::<BitsAttributeArgs>(attr.tts.clone().into()).unwrap();
+                    let expected_bits = bits_arg.size;
+                    bits_check_tokens.extend(quote_spanned! { expected_bits.span() =>
+                        let _ = bitfield::checks::BitsCheck::<
+                            [(); #expected_bits]
+                        >{
+                            arr: [(); <#field_type as bitfield::Specifier>::BITS]
+                        };
+                    })
+                }
+
+                tokens.extend(quote!{
+                    pub fn #getter_name(&self) -> <#field_type as bitfield::Specifier>::Face {
+                        #bits_check_tokens
+
+                        <#field_type as bitfield::Specifier>::Face::from_bits(
+                            bitfield::Bits(self.get::<#field_type>(#offset))
+                        )
+                    }
+
+                    pub fn #setter_name(&mut self, new_value: <#field_type as bitfield::Specifier>::Face) {
+                        self.set::<#field_type>(#offset, new_value.into_bits().into_raw())
+                    }
+                });
+                offset.push(syn::parse_quote!{ <#field_type as bitfield::Specifier>::BITS });
+            }
+        } else {
+            unreachable!()
+        }
+    }
+
+    pub fn has_generics(&self) -> bool {
+        self.ast.generics.lt_token.is_some() && self.ast.generics.gt_token.is_some()
+    }
+
+    pub fn validate(&self) -> ParseResult<()> {
+        if self.has_generics() {
+            return bail!(
```

Comment:
The bail macro already does a return, so the return on this line is unreachable.